src/renderer/utils/createStorage.ts

```diff
@@ -1,73 +1,210 @@
+      ...options,
+    });
+    this.data.observe(change => {
+      const { newValue, oldValue } = toJS(change, { recurseEverything: true });
+
+      this.onChange(newValue, oldValue);
+    });
+  }
 
-    return this.set(Object.assign(currentValue, value));
+  protected onChange(value: T, oldValue?: T) {
+    if (!this.initialized) return;
+
+    this.storage.onChange?.({ value, oldValue, key: this.key });
+    this.storage.setItem(this.key, value);
   }
 
-  clear() {
-    this.storage.removeItem(this.key);
+  load(): T {
+    return this.storage.getItem(this.key);
+  }
 
-    return this;
+  get(): T {
+    return this.data.get();
   }
 
-  getDefaultValue() {
-    return this.defaultValue;
+  set(value: T) {
+    this.data.set(value);
+  }
+
+  reset() {
+    this.set(this.defaultValue);
+  }
+
+  clear() {
+    this.data.set(null);
   }
 
-  restoreDefaultValue() {
-    return this.set(this.defaultValue);
+  merge(value: Partial<T> | ((draft: Draft<T>) => Partial<T> | void)) {
+    const updater = isFunction(value) ? value : (state: Draft<T>) => {
+      if (isObject(state)) return Object.assign(state, value);
+
+      return value;
+    };
+    const currentValue = this.toJS();
+    const nextValue = produce(currentValue, updater) as T;
+
+    this.set(nextValue);
+  }
+
+  toJS() {
+    return toJS(this.get(), { recurseEverything: true });
   }
 }
+
+export const localStorageAdapter: StorageAdapter<object> = {
+  getItem(key: string) {
+    return JSON.parse(localStorage.getItem(key));
+  },
+  setItem(key: string, value: any) {
+    localStorage.setItem(key, JSON.stringify(value));
+  },
+  removeItem(key: string) {
+    localStorage.removeItem(key);
+  }
+};
+
+// TODO: remove after merge https://github.com/lensapp/lens/pull/2269
+export const jsonFileSyncStorageAdapter: StorageAdapter<object> = {
+  cwd: path.resolve((app || remote.app).getPath("userData"), "lens-local-storage"),
+  data: observable.map<string, any>([], { deep: false }),
+  initialized: false,
+
+  get filePath() {
+    const clusterId = location.hostname.split(".").slice(-2, -1)[0];
+
+    return path.resolve(this.cwd, `${clusterId ?? "app"}.json`);
+  },
+
+  init() {
+    if (this.initialized) return;
+
+    try {
+      ensureDirSync(this.cwd, { mode: 0o755 });
+      const data = readJsonSync(this.filePath, { throws: false }) || {};
+
+      this.data.replace(data);
+      this.bindAutoSave();
+    } catch (error) {
+      console.error(`[init]: ${this.filePath} failed: ${error}`, this);
+    } finally {
+      this.initialized = true;
+    }
+  },
+
+  bindAutoSave() {
+    return reaction(() => this.data.toJSON(), async (data) => {
+      try {
+        await writeJson(this.filePath, data, { spaces: 2 });
+      } catch (error) {
+        console.error(`[save]: ${this.filePath} failed: ${error}`, this);
+      }
+    });
+  },
+
+  getItem(key: string) {
+    this.init();
+
+    return this.data.get(key);
+  },
+  setItem(key: string, value: any) {
```

Comment:
We shouldn't have this sort of coercion in a storage abstraction.
```suggestion
  setItem(key: string, value: object) {
```